### PR TITLE
Don't save parents unless needed

### PR DESCRIPTION
**Pull request by cheeetoo**
Speeds up operations that do not require gradients by a bit since it no longer saves parents, and also speeds up backward pass since there is a less expensive check for if backward should be run.

### CODE REVIEW

**Comment by wozeparrot** [comment]
tests? does it provide a greater than 20% speedup?

**Comment by cheeetoo** [comment]
It does not provide a >20% speedup, is that a requirement? I am not sure what you mean by tests, sorry.

**Comment by wozeparrot** [comment]
> It does not provide a >20% speedup, is that a requirement? I am not sure what you mean by tests, sorry.

I'm not sure if this will hit any edge cases, and I don't think we have great test coverage around stuff like this, I stand to be corrected tho.

**Comment by geohot** [comment]
Hmm, actually, I think this is good idea. The main thing it could save is RAM in freeing the parents.

**Comment by roelofvandijk** [comment]
> Hmm, actually, I think this is good idea. The main thing it could save is RAM in freeing the parents.
> 
> Though it's sort of crazy this wasn't already done. The test coverage is the whole tinygrad...everywhere we do a backward pass.
> 
> I'm racking my brain about why I wrote `any(x.requires_grad for x in t0._ctx.parents)` and not just `t0.requires_grad`

If you can figure this out, there's quite the speedup hiding there. These __inits__ are called a LOT.
